src/tests/auth/test_auth_gitlab.py

```diff
@@ -0,0 +1,202 @@
+import copy
+import json
+import os
+import tempfile
+import time
+import unittest
+
+from tornado import escape
+
+from auth.auth_gitlab import GitlabOAuthAuthenticator
+from model import server_conf
+from tests import test_utils
+from utils import file_utils
+from unittest import TestCase
+from unittest.mock import patch, Mock
+
+if __name__ == '__main__':
+    unittest.main()
+
+mock_time = Mock()
+mock_time.return_value = 10000.01
+mock_dump_sessions_to_file = Mock()
+mock_do_update_groups = Mock()
+mock_do_update_user = Mock()
+mock_request_handler = Mock(**{'get_secure_cookie.return_value': "12345".encode()})
+
+
+class TestAuthConfig(TestCase):
+    @patch('time.time', mock_time)
+    @patch('auth.auth_gitlab.GitlabOAuthAuthenticator.dump_sessions_to_file', mock_dump_sessions_to_file)
+    @patch('auth.auth_gitlab.GitlabOAuthAuthenticator.do_update_groups', mock_do_update_groups)
+    def test_gitlab_oauth(self):
+        tmp = tempfile.mkstemp('.json', 'test_auth_gitlab-')
+        now = time.time()
+        state = {
+            "user@test.com": {
+                "groups": ["testgroup"],
+                "updating": False,
+                "updated": now-10,
+                "visit": now-10,
+                "id": 1,
+                "username": "test",
+                "name": "John",
+                "email": "user@test.com",
+                "state": "active"
+            },
+            "nogroups@test.com": {
+                "groups": None,
+                "updating": True,
+                "updated": now-10,
+                "visit": now-10,
+                "id": 2,
+                "username": "nogroups",
+                "name": "John",
+                "email": "nogroups@test.com",
+                "state": "blocked"
+            }
+        }
+
+        os.write(tmp[0], str.encode(escape.json_encode(state)))
```

Comment:
Maybe you could use test_utils here? There is `test_utils.temp_folder` and `test_utils.create_file`

Also in the test class you do 
```
    def setUp(self):
        test_utils.setup()

    def tearDown(self):
        test_utils.cleanup()
```
And in this case the data will always be cleaned up after execution. At the moment, if any assertion fails, you won't close the temp file.

Reply:
Ok